src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java

```diff
@@ -64,39 +71,43 @@ public Rectangle getArea() {
 
             @Override
             public void accept(@Nonnull Object ingredient) {
-                FluidStack stack;
-                if (ingredient instanceof FluidStack) {
-                    stack = (FluidStack) ingredient;
-                } else {
-                    stack = drainFrom(ingredient);
-                }
+                FluidStack stack = getFluidFromIngredient(ingredient);
 
                 if (stack != null) {
                     NBTTagCompound compound = stack.writeToNBT(new NBTTagCompound());
                     writeClientAction(LOAD_PHANTOM_FLUID_STACK_FROM_NBT, buf -> buf.writeCompoundTag(compound));
                 }
 
-                if (isClient) {
-                    phantomTank.setFluid(stack);
-                }
+                phantomFluidSetter.accept(stack);
             }
         });
     }
 
+    @Nullable
+    private static FluidStack getFluidFromIngredient(Object ingredient) {
+        if (ingredient instanceof FluidStack) {
+            return (FluidStack) ingredient;
+        } else if (ingredient instanceof ItemStack) {
+            ItemStack itemStack = (ItemStack) ingredient;
+            IFluidHandlerItem fluidHandler = itemStack.getCapability(CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY, null);
+            if (fluidHandler != null)
+                return fluidHandler.drain(Integer.MAX_VALUE, false);
+        }
+        return null;
```

Comment:
This should probably throw an exception instead of returning null. Passing something else into this method ought to be treated as a critical failure.

Reply:
I think hard crashing on that spot would make the game crash whenever you drag any wrong ingredient into the slot.